target/linux/lantiq/patches-5.15/0751-net-dsa-lantiq_gswip-allow-adding-fdb-for-individual.patch

```diff
@@ -0,0 +1,40 @@
+From a115138421020267291668e7a0937be96197110a Mon Sep 17 00:00:00 2001
+From: Hauke Mehrtens <hauke@hauke-m.de>
+Date: Sun, 30 Jul 2023 23:38:25 +0200
+Subject: net: dsa: lantiq_gswip: allow adding fdb for individual ports
+
+This allows adding a MAC address into the forwarding database (fdb) for
+ports which are not part of a bridge. This is happening since kernel
+5.15 and the driver rejected it previously with an error message.
+
+Signed-off-by: Hauke Mehrtens <hauke@hauke-m.de>
+---
+ drivers/net/dsa/lantiq_gswip.c | 16 +++++++++-------
+ 1 file changed, 9 insertions(+), 7 deletions(-)
+
+--- a/drivers/net/dsa/lantiq_gswip.c
++++ b/drivers/net/dsa/lantiq_gswip.c
+@@ -1445,14 +1445,16 @@ static int gswip_port_fdb(struct dsa_swi
+ 	int i;
+ 	int err;
+ 
+-	if (!bridge)
+-		return -EINVAL;
+-
+-	for (i = max_ports; i < ARRAY_SIZE(priv->vlans); i++) {
+-		if (priv->vlans[i].bridge == bridge) {
+-			fid = priv->vlans[i].fid;
+-			break;
++	if (bridge) {
++		for (i = max_ports; i < ARRAY_SIZE(priv->vlans); i++) {
++			if (priv->vlans[i].bridge == bridge) {
++				fid = priv->vlans[i].fid;
++				break;
++			}
+ 		}
++	} else {
++		/* FID of a standalone port (single port bridge) */
++		fid = port + 1;
```

Comment:
This patch is better, but it needs some extra features introduced between 5.15 and 6.1:
https://github.com/xdarklight/linux/commit/4aadb97a77807ef954c49c4e47369dbb5ebcadef

Reply:
Unfortunately this is not working. Yes, it silences the warnings that people have been seeing.
But that doesn't meant that it makes the GSWIP driver comply with DSA's expectations.

It's been a while but from memory the issue is that DSA expects that it can add/remove static FDB entries for a port at any time.
Unfortunately this is tricky to implement for the GSWIP driver. 
A as long as a port is a standalone port (meaning it's not part of a bridge) the `fid = port + 1` logic is correct.
As soon as the port is part of a bridge this is not correct anymore a different fid (the one assigned to the bridge) has to be used.
But there's cases where a static FDB entry is added early before the port is part of a bridge. Then the port is added to the bridge, meaning it's fid changes. This is not reflected in the static FDB entries and so basically what happens is that any per port FDB entries that were added before the port became part of a bridge are not used by the hardware anymore.

Reply:
I also forgot: this does not honor `DSA_DB_PORT` yet.
We should move to something like https://github.com/xdarklight/linux/commit/4aadb97a77807ef954c49c4e47369dbb5ebcadef - but again, even that still needs (at lot of) work

Reply:
`DSA_DB_PORT` is not available in kernel 5.15. It was added with kernel 5.18, see:
https://github.com/torvalds/linux/commit/c26933639b5402c174c65c01d33f145622784012

Maybe we can migrate this forwarding rule when we add the port to a bridge?